core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -749,10 +753,10 @@ private boolean allocateUnassigned() {
                     } else if (logger.isTraceEnabled()) {
                         logger.trace("No Node found to assign shard [{}]", shard);
                     }
-                    unassigned.ignoreShard(shard);
+                    changed |= unassigned.ignoreShard(shard, Optional.ofNullable(decision));
```

Comment:
On second look, throttledNodes only contains NO decisions, as deciders.canAllocate(node, allocation) can only return YES or NO (there is just one decider that implements this method). We can thus assert that when we add to throttledNodes, that the decision was NO. We can thus conclude that at this line, the decision is NO.